### PR TITLE
Small fixes for --crate-type staticlib

### PR DESCRIPTION
**Pull request by bjorn3**
The first commit doesn't have an effect until we start translating error messages. The second commit fixes potential linker errors when combining `--crate-type staticlib` with another crate type and I think `-Cprefer-dynamic`.

### CODE REVIEW

**Review by estebank — APPROVED** [review]

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks @bjorn3!

@bors r+

**Comment by bjorn3** [comment]
I didn't account for it being possible for a crate to have 0 crate types like would happen if you use `--crate-type cdylib` on a platform not supporting cdylib's.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -230,6 +230,9 @@ pub fn each_linked_rlib(
                 });
             }
         }
+        if info.dependency_formats.is_empty() {
+            return Err(errors::LinkRlibError::MissingFormat);
```
Comment (wesleywiser) [inline compiler/rustc_codegen_ssa/src/back/link.rs:234]: Won't this still result in a compilation error/failed test case for that target as it eventually leads to

https://github.com/rust-lang/rust/blob/77051167083b58c4cd59c658197bcfe6476a7fa5/compiler/rustc_codegen_ssa/src/back/link.rs#L581

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/link.rs:234]: This branch is only reachable when no crate type is passed to `each_linked_rlib`, which only happens during the LTO code. This code explicitly ignores the error. `link_staticlib` passes in `CrateType::StaticLib` and in addition is only called if `--crate-type staticlib` is used, in which case `dependency_formats` has an entry for `CrateType::StaticLib` and as such is not empty. In addition this code matches what happened before this PR.

Comment (wesleywiser) [reply compiler/rustc_codegen_ssa/src/back/link.rs:234]: Thanks!

**Pull request merged**